tiledb/sm/c_api/tiledb.h

```diff
@@ -3201,6 +3234,468 @@ TILEDB_EXPORT int32_t tiledb_array_encryption_type(
+ * @endcode
+ *
+ * @param ctx The TileDB context.
+ * @param subarray The subarray.
+ * @param domain The domain to be retrieved.
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT int32_t tiledb_subarray_get_domain(
+    tiledb_ctx_t* ctx, const tiledb_subarray_t* subarray, const void** domain);
+
+/**
+ * Adds a 1D range along a subarray dimension. The datatype of the range
+ * must be the same as the subarray domain type.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * uint32_t dim_idx = 2;
+ * int64_t range[] = { 10, 20 };
+ * tiledb_subarray_add_range(ctx, subarray, dim_idx, range);
+ * @endcode
+ *
+ * @param ctx The TileDB context.
+ * @param subarray The subarray to add the range to.
+ * @param dim_idx The index of the dimension to add the range to.
+ * @param range The range to add.
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT int32_t tiledb_subarray_add_range(
+    tiledb_ctx_t* ctx,
+    tiledb_subarray_t* subarray,
+    uint32_t dim_idx,
+    const void* range);
+
+/**
+ * Retrieves the number of ranges of the subarray along a given dimension.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * uint64_t range_num;
+ * tiledb_subarray_get_range_num(ctx, subarray, dim_idx, &range_num);
+ * @endcode
+ *
+ * @param ctx The TileDB context
+ * @param subarray The subarray.
+ * @param dim_idx The index of the dimension whose range number to retrieve.
+ * @param range_num The number of ranges to retrieve.
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT int32_t tiledb_subarray_get_range_num(
+    tiledb_ctx_t* ctx,
+    const tiledb_subarray_t* subarray,
+    uint32_t dim_idx,
+    uint64_t* range_num);
+
+/**
+ * Retrieves the number of ranges of the subarray along a given dimension.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * const void* range;
+ * tiledb_subarray_get_range(ctx, subarray, dim_idx, range_idx, &range);
+ * @endcode
+ *
+ * @param ctx The TileDB context
+ * @param subarray The subarray.
+ * @param dim_idx The index of the dimension to retrieve the range from.
+ * @param range_idx The index of the range to retrieve.
+ * @param range The range to retrieve.
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT int32_t tiledb_subarray_get_range(
+    tiledb_ctx_t* ctx,
+    const tiledb_subarray_t* subarray,
+    uint32_t dim_idx,
+    uint64_t range_idx,
+    const void** range);
+
+/**
+ * Retrieves the estimated result size for a fixed-sized attribute.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * uint64_t size;
+ * tiledb_subarray_get_est_result_size(ctx, subarray, "a", &size);
+ * @endcode
+ *
+ * @param ctx The TileDB context
+ * @param subarray The subarray.
+ * @param attr_name The attribute name.
+ * @param size The size (in bytes) to be retrieved.
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT int32_t tiledb_subarray_get_est_result_size(
```

Comment:
I recall some confusion around this -- can there be some additional explanation in the docstring of how users should use the estimate provided by this API?

Reply:
We will address any C API issues in a subsequent PR.